### DISCUSSION TITLE
Using Rockpi 4c+ with LTSP

### DISCUSSION THREAD

**Discussion by jtornero** [ltsp/ltsp#829] [General]
Well, we use LTSP for a onboard survey system which has several sampling kiosks. Those sampling kiosks are based on a nice mixtures of Raspberry Pi computers (2B+, 3B+, 3B, etc). The system works like a charm, 

**BUT**

unfortunately Raspberry Pi devices are nowadays a pain to get: At least in Spain, they are VERY expensive (current price is about 200 €) and even worse, they are completely out of stock everywhere. 

I've received @alkisg advice about moving to miniPC's or whatever, providing they are way more easy to operate. However, I work for an R&D government agency and trying to get such devices would bury me in red tape until 2027 at least, hahahhaha.
So the logical move is to replace Raspberry Pi for another board. We have choosen Rock 4C+ because it is easy to get int his moment and also there was some information about the chances of network boot.

However, is not as easy as with the Raspberry Pi but I'd like to share with you what I've done so far, so maybe someone can make use it and, most important, someone can help me a little.

First of all: Rock 4C+ boards are able to do PXE boot. Unfortunately, the available distributions for these boards don't have a boot menu but you can access the u-boot menu while booting, set server and board ip, and get PXE configuration and boot it. So the main difficulty was to compile u-boot with de appropiate code to make the PXE boot. This is, at least for me, an absolute ordeal.

However one of the official OS images (DEBIAN desktop) provides a way to modify the boot options, making possible to adjust things in the file `extlinux.conf` and reconfiguring the kernel for the changes to work.

Providing the things that are neccessary for PXE boot are already met by the LTSP installation (DHCP server, TFTP server, NFS server) it was reasonably easy to make it work: Just make a folder under `/srv/ltsp` for the operating system files, which I have copied from the img file provided to work, basically the same thing that we do with Raspberry Pi following the LTSP site tutorial.

**There is a little trick here:** The board needs a SD card for booting. And the original Debian Image doesn't have the PXE boot implemented, so what I have done is:

- Downloaded the image for the OS.
- Installed it in a SD card and run for the first time the OS (this creates an user in the SD card), this is more or less that doing it for a Raspberry
- I get root access in the booted system and do the modifications for PXE boot, basically providing `nfsroot` and stuff. This can be done modifyng the file `/etc/kernel/cmdline` and running a script called `u-boot-update` on the client that makes the magic to update u-boot etc.
- Now I losetup, mount and copy the debian image in `/srv/ltsp/rock4cplus` in the server and also copy the contents of `/boot `int tftp directory (this is made "by hand")
- Then I am able to chroot into the client, install packages (LTSP).

My installations is as follows:

**SERVER**
LTSP version: 21.01-1~ubuntu20.04.1
SERVER XSESSIONS: ubuntu.desktop
SERVER OS: PRETTY_NAME="Ubuntu 20.04.2 LTS"
CHROOTS:
raspios
rock4cplus
VMs:
IMAGES:
raspios
rock4cplus

**ROCK4CPLUS CLIENT**
LTSP version: 21.01-1
SERVER XSESSIONS: openbox.desktop  plasma.desktop
SERVER OS: PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
CHROOTS:

VMs:

IMAGES:

Of course I do _ltsp ipxe, ltsp nfs, ltsp initrd_, etc.
By the way my server _/etc/ltsp/ltsp.conf_ is [here](https://gist.github.com/jtornero/11da42c7828735dceaa20007aedd6937) and my NFS exports are

```
root@pequepel2:~# exportfs
/srv/ltsp       <world>
/srv/tftp/ltsp  <world>
/srv/ltsp/raspios
                <world>
/srv/ltsp/rock4cplus
                <world>
```

Then I can boot it using LTSP in what I guess is NFS_RW mode (because I am able to install things etc and make it persistent). Nothing special to do, just plug it and wait. [The boot log is here](https://gist.github.com/jtornero/3120a75655d5f60ad0c210e500f0d783) 

**HOWEVER**

1.  I can't get LTSP mode working. Modifying `/etc/kernel cmdline` with the init and ltsp.image code provided for Raspberry Pi `cmdline.txt` for LTSP mode [leads to a boot interruption](https://gist.github.com/jtornero/3d237acd73aced3536a75c086e6aea19) (Unfortunately I am not able to properly enable the DEBUG SHELL).
2. Neither I am not able of mounting the home directories in the server (Well, that's part of LTSP mode, isn't it?)
3. I notice that the work on the home folder of the booted LTSP mode Rock 4C+, i.e. saving a txt file, is written in the SERVER `/home/radxa` folder (luckily it is consistent with NFS_RW mode)

So the questions are: How to keep going on? What could cause the problem trying to boot in LTSP mode?

By the way, my LTSP setup was running with Raspberries and after all that mess the system is still working with the Raspberries, so at least one thing is more than perfect!!!!

Thank you very much

JOrge

**Comment by danboid** [comment]
I don't have an answer as I've never done what you're trying to do before but I do want to add my 2p here:

It looks like (I'm guessing) you're using a RPi 4 running Ubuntu 20.04 as your LTSP server? If you want to use Rock Pi4 SBCs as your clients then you will have an easier time use a RockPi 4 as a server too, that way you can run LTSP chrootless. Chrootless LTSP servers are the easiest to maintain and also the most likely to work.

I'd run Ubuntu 22.04 on your server, if you prefer Ubuntu over Debian, whether its a RPi or a RockPi or x86. If you need to stick with Ubuntu 20.04 then @alkisg is probably going to recommend you install LTSP from git instead of using an old LTSP release.

The LTSP docs don't seem to caution people from using chroots but I had problems with trying to use chroots. Either use chrootless or create your LTSP images in a VM - thats tricky when you're using ARM though so chrootless is your only sensible option, at least at first, IMO. I use the VM option for my x86 LTSP server because that way you get the most control over your client image but when starting off, chrootless.

@alkisg Does the LTSP PPA work with aarch64 Ubuntu too?

**Reply by jtornero** [reply]
Well... good news

I deleted some commands from the cmdline (I left only one `console={whatever}` command and also removed `plymouth.ignore-serial-consoles`) and refreshed the kernel etc. in the SD card boot hoping It would help to see the error (I wasn't able to activate LTSP DEBUG_CONSOLE) This was a desperate trial-and-error but...

Finally I saw the error. Unfortunately I forgot to save the boot log, but it was clear that it was a problem of mounting nfs shares.

At the end, it looks like RockPi 4C+ Debian official image is not shipped with `nfs-common` package. I installed it into the chroot AND IT WORKS!!!!

I am able to log in with any of the users in the server and the /home folder is mounted from the server, not from the LTSP image

To mantain I guess I should be more proffesional but with a chroot /srv/ltsp/rockpi works perfectly for installing packages, software etc.

I'd like to make all the installation again and create a sort of guide for RockPi 4C+. **Spolier**: At the emd, it is almost as esay as setting it up for Raspberry Pi.  The @alkisg guide is almost valid for Rockpi, most differences are relative to messing with the cmdline kernel stuff and doing the trick for enabling PXE boot from the Rockpi the first time.

Maybe you can try with your Rockpi 4

All the best

Jorge

**Reply by danboid** [reply]
Great news Jorge!

What image are you booting on your RockPi LTSP clients? What OS and is it a chroot, chrootless or...

A more detailed guide for the LTSP wiki would be great.

**Reply by jtornero** [reply]
Hi again,

I boot the Official Rock 4C+ Debian image available [here](https://github.com/radxa-build/rock-4c-plus/releases/download/20230312-1521/rock-4c-plus_debian_bullseye_kde_b55.img.xz)

The installation is in a chroot. LTSP server serves this image and also a Raspberry Pi image

For the guide we have to wait  little, but I guess I'll manage to get it soon. I want to make a completely fresh install of LTSP and the images etc. because I've been messing a little with the /srv/ltsp files etc. 

The worst problem I have is I'm sorry but I'm a dilettante and some things I've done I don't understand completely but at the end it works. I hope someone could just check if there is something wrong or if I missed any important step.

All the best

Jorge

**Reply by danboid** [reply]
I'm a bit out of touch with the Rock Pi 4 as I've not used mine in a while but you are aware that Armbian support the RockPi 4 and offer both Ubuntu and Debian based images, which you could potentially use for both the LTSP server or clients? It could be that the official images work better for you but its an option I'd recommend investigating if you've not done so already.

https://www.armbian.com/rockpi4/

**Reply by jtornero** [reply]
Well I found a problem with many of those images. Not sure how to explain, because it is far beyond from my knowledge, but anyway it looks like one of the problems is that the images organize their partitions in different ways:

- Debian image is 2 partitions, one is the / partition (with /boot inside) ant the other is a config partition which stores two files for first boot and it looks that it is ok with LTSP
- Another images are split into 5 partitions and just say I wasn't able to make them boot, I had to mess a lot for getting kernel and stuff files. And also not all partitions can be mounted.

But apart from this the most important thing here is to be able to modify uboot to make it possible a PXE boot. And in this case I have been able to do it only with Debian image. It is possible to compile a Armbian image and fiddle with uboot but is fairly complex and, worse, last week I was totally unable to compile because of some broken things in 3rd party sites and the compilation tool.

Moreover, right now Rock 4C is "Platinum supported"  with Armbian but Rock 4C**+** isn't as well supported and in fact in the Armbian compilation script is still considered experimental.

So despite it would probably be easy to boot another of the images by an experienced person... I'm sorry I don't have the necessary skills to do that :(

However, right now I'm making a fresh LTSP install and I think in short I'll be able to write all the steps down.

Cheers

**Reply by danboid** [reply]
That's a shame re Armbian's 4c+ support but it doesn't affect me as I've got a normal 4c.

Thanks for documenting it!

**Comment by jtornero** [comment]
Well, here is a preliminar guide

### Using Rock 4C+ boards with LTSP

## Server installation

I started with a fresh Ubuntu 20.04 LTS server install. This choice was made because the server is an old HP xw8400 workstation with a RAID and after many trials I was unable to run the 22.04 LTS installer. However, at least for the purposes of this guide it doesn't matter.

For LTSP installation [I followed the steps in LTSP site](https://ltsp.org/docs/installation/), but without all the epoptes stuff.

`apt install --install-recommends ltsp ltsp-binaries dnsmasq nfs-kernel-server openssh-server squashfs-tools ethtool net-tools`

Because we have 2 NIC's in our server, I had to run 

`ltsp-dnsmasq --proxy-dhcp=0`

and also, after identifying each of the NIC's, set up the network properly. 

Also I had to edit `/etc/dnsmasq.d/ltsp-dnsmasq.conf` so I ensure the DHCP server only listens to the LTSP network interface, the range of the IP for dhcp is properly set for our needs and the DNS server is set properly:

```
interface=ens6
dhcp-range=192.168.2.150,192.168.2.250,12h
```
It may be necessary to set another options depending on your network setup.

Then we do
 ```
ltsp ipxe
ltsp nfs
ltsp initrd
```

This will create, among other stuff, the folders `/srv/ltsp` y /`srv/tftp` in our server.

Now with `/srv/ltsp` available we can setup our operating system image.

## Creation of operating system image in the server (Preparing the chroot)

First of all I have to mention that there are several OS available for Rock 4C+. However, after many trial-and-error tests, **AFAIK that only the official Rock 4C+ Debian image makes it possible**. The reasons are:

- The image is divided into 2 partitions, one for first setup scripts of boards (/config) and the other for all the / filesystem. Other images are divided in 5 partitions and it was impossible (for me) to make it work.
- The main reason, however, is that most of the tested images are not able to do a PXE boot unless you enter uboot menu on every boot and issue the appropiate uboot commands. Moreover, the uboot implementation in these images makes not possible to save the environment in a SD card.
Fortunately the official Rock 4C+ Debian image provides a reasonably simple way to get this done. We will get back on this later when preparing the SD card for booting.

- Now we are going to follow most of the [steps given in ltsp.org for Raspios](https://ltsp.org/docs/installation/raspios/) but adapted to our needs:

1. [Download the official Rock 4C+ Debian image](https://github.com/radxa-build/rock-4c-plus/releases/download/20230312-1521/rock-4c-plus_debian_bullseye_kde_b55.img.xz), decompress the downloaded file and if you want, rename the obtained `img` file to a simpler name to ease the things later, let's say `rock4cplus.img`.

2. Then we setup a loopback device with the downloaded image:
```
losetup -f # to get a free device, we will assume /dev/loop8
losetup -rP /dev/loop8 rock4cplus.img
```

If we make ls /dev we notice that two new partitions will appear, in my case they were `/dev/sdd1` and `/dev/sdd2.` You must figure out which are the names of yours, anyway you must mount the root filesystem partition in the next step. In my case it was `/dev/sdd2.`

3. We mount it and copy the files in the root filesystem to our /srv/ltsp folder:
`cp -a /mnt/. /srv/ltsp/rock4cplus`

4. After that, we unmount the filesystem and detach the loopback device.
```
umount /mnt
losetup -d /dev/loop8
```
Now we have our chroot under /srv/ltsp/rock4cplus, but we still have things to do to be able to use it properly.

5. Because the different architectures involved, you'll need to install qemu-user-static to be able to chroot:
`apt-get-install qemu-user-static`

6. We have to be able to resolve names, etc in our chroot, so we need to do this (To be better documented, help here would be very appreciated)
```
sudo mount -o bind /run /mnt/run
echo 'nameserver 8.8.4.4' | sudo tee -a /etc/resolv.conf
```
Now we are able to `chroot /srv/ltsp/rock4cplus`

7. We must install LTSP packages into the client's chroot:
```
apt-get update
apt install --install-recommends ltsp
```
8. **VERY IMPORTANT** the Debian image is not provided with the package nfs-common by default. We need to install it to be able to mount all the NFS partitions in the clients.
`apt install nfs-common`

9. After this we exit from the chroot (just send `exit` in the chroot prompt) and send some LTSP commands to make it work:
```
ltsp kernel
ltsp ipxe
ltsp nfs
ltsp initrd
```
10. Now it's just matter of creating the compressed rock4cplus image:

`ltsp image rock4cplus --mksquashfs-params='-comp lzo'`

## Modifying the OS image to boot from network

This has been made mostly by trial-and-error so better protocols for doing it could exists for sure. As  I said before, only the Debian image is capable of modify kernel/uboot parameters to make possible PXE boot. To shorten we need to follow these steps:

1. Write the Debian Image in a SD card: You can do this with many softwares out there, however I got better results with Startup Disk Creator.

2. Then comes the tricky part: to update the kernel arguments so uboot can do PXE boot. Because we are able to mantain the server filesystem with chroot, we can set it up for work just in LTSP mode. This means that the booted cards couldn't modify the operating system and also that the user folder will be those of the server.

I reccomend you do all these steps through the serial console. [You can checkout how to do it here](https://wiki.radxa.com/Rockpi4/dev/serial-console) 
Because serial console speed is the not-very-standard 1500000, you must have a USB to RS232 adapter capable of managing such speeds. After some trials [I found this inexpensive device based in the chip CH340G that works like a charm.](https://www.amazon.es/dp/B08T22HC8K?psc=1&ref=ppx_pop_dt_b_product_details)

1. Once the image is written, boot one of your boards with it. 
2. Log in and gain sudo/root access (User is `radxa`,  password `radxa`)

```
cd /boot
```
4. Inside the boot folder there is a text file, uEnv.txt with the following content:
```
# uEnv.txt has been retired.
# Please use:
#   'rsetup' command to confiure device overlays;
#   'sudo nano /etc/kernel/cmdline && sudo u-boot-update' to update kernel arguments.
```
We will modify cmdline [adding mostly the same parameters as stated in Raspbios](https://ltsp.org/docs/installation/raspios/#ltsp-mode-netbooting) installation.

So just do `sudo nano /etc/kernel/cmdline && sudo u-boot-update` and modify cmdlin file that should be left as follows (notice there is a single line:

```
loglevel=4 ip=dhcp root=/dev/nfs nfsroot=192.168.2.1:/srv/ltsp/rock4cplus,vers=3,tcp,nolock init=/usr/share/ltsp/client/init/init ltsp.image=images/rock4cplus.img console=ttyS2,1500000n8 console=ttyFIQ0,1500000n8 coherent_pool=2M irqchip.gicv3_pseudo_nmi=0 quiet splash
```

**IMPORTANT:** Remember to match nfsroot server IP and path for those of your particular setup (by default, LTSP uses 192.168.67.1) and also the name of the image.

`Modifyng `/etc/kernel/cmdline` and doing `u-boot-update` will update kernel arguments and also modifiy another of our targets, the file `extlinux.conf` in `/boot/extlinux` . This file is where the uboot menu lives and with this steps done it is enough to boot the boards with the OS inside the SD card. However, it is desirable to take some further steps, because in this state the board will need one particular file, the Device Tree descriptor file, to be able to boot properly. With this modification we will be able to reduce a lot the size of the SD card necesary to boot the boards and also simplify it. The idea is to reduce the SD card filesystem as much as possible, but the board need three files from the SD card to make the first booting steps:
- initrd.img
- vmlinuz
- and the device tree file, in our case `rk3399-rock-4c-plus.dtb`

The first two files are located at SD card /boot but the third is located outside /boot, at /usr/lib/linux-image-5.10.110-1-rockchip/
So if we want to reduce the filesystem the best thing to do is to remove all the unnecesary files from the SD card.
So what we do is copy the dtb file in /boot. This can be done also mounting the SD card in an appropiate mount point of the server, in that case you'll have to provide the appropiate paths.

```
mkdir /boot/fdt
cp  /usr/lib/linux-image-5.10.110-1-rockchip/rockchip/rk3399-rock-4c-plus.dtb /boot/fdt
```
But still we need to tell uboot where to find the dtb file, because every time we mess editing `cmdline` and then `u-boot-update` the `extlinux.conf` file is overwritten and a default path for kernel, initrd, etc is set. So we will just manually edit` /boot/extlinux/extlinux.conf` providing the proper path for the dtb file. This can be done modifyng the line `fdtdir` in the appropiate section of the file (in our case, the commands under l0, the default label)

Instead of creating an specific folder for the dtb file you can just just in /boot, anyway, just make sure that you provide the correct path for `fdtdir` in the previous step. Also you should remove the `root={A NICE UUID}` command after `append`.

 Your extlinux.conf file should look more or less like this:

```
default l0
menu title U-Boot menu
prompt 0
timeout 10

label l0
        menu label LTSP boot
        linux /boot/vmlinuz-5.10.110-1-rockchip
        initrd /boot/initrd.img-5.10.110-1-rockchip
        fdtdir /boot/fdt/

        append ip=dhcp root=/dev/nfs nfsroot=192.168.2.1:/srv/ltsp/rock4cplus,vers=3,tcp,nolock init=/usr/share/ltsp/client/init/init ltsp.image=images/rock4cplus.img loglevel=4 console=ttyS2,1500000n8 console=ttyFIQ0,1500000n8 coherent_pool=2M irqchip.gicv3_pseudo_nmi=0 quiet splash

label l0r
        menu label Debian GNU/Linux 11 (bullseye) 5.10.110-1-rockchip (rescue target)
        linux /boot/vmlinuz-5.10.110-1-rockchip
        initrd /boot/initrd.img-5.10.110-1-rockchip
        fdtdir /usr/lib/linux-image-5.10.110-1-rockchip/

        append root=UUID=8087ba46-93d3-4f58-a640-231fd0aa806c ip=dhcp root=/dev/nfs nfsroot=192.168.2.1:/srv/ltsp/rock4cplus,vers=3,tcp,nolock init=/usr/share/ltsp/client/init/init ltsp.image=images/rock4cplus.img loglevel=4 console=ttyS2,1500000n8 console=ttyFIQ0,1500000n8 coherent_pool=2M irqchip.gicv3_pseudo_nmi=0 splash single
```
After this step it is safe to delete all files and folders in the SD card except of /boot folder. Then you can take your SD card and enjoy booting your Rock 4C+ boards into LTSP mode.

### Shrinking clients SD image file

The SD card we have just prepared can be cloned for booting more clients. However, despite having the minimum possible set of files for boot it is still big: more than 5 GB, because we have deleted files but the filesystem has not been reduced.

Looking around[ I found this ](https://askubuntu.com/questions/1174487/re-size-the-img-for-smaller-sd-card-how-to-shrink-a-bootable-sd-card-image) where the process is well explained.

As a summary, you'll have to follow these steps:

1. With dd, make a dump of the SD card to a img file
```
sudo dd if=/dev/sdd of=SD.img bs=512
```
2. Find a free loopback device and create a device from the image file
```
sudo losetup -f
losetup /dev/loop7 SD.img
```
3. Because we need to access the partitions we must tell the kernel to do so
```
sudo partprobe /dev/loop7
```
4. Now with `gparted` we resize our partition to the minimum possible space. To do so select the device in gparted, then select the partition and resize it (Sorry, no pictures of it)
 
5. Detach the device with `sudo losetup -d /dev/loop7` and then you'll have the SD.img file with the partition reduced.

Now comes the tricky part. We have reduced the partitions but the image file is not reduced yet (Remember that you left a lot of unallocated space when used gparted??) so it's time to _shave_ the img file. This is tricky because of the type of partition table, fortunately there is a comment to the stackoverflow answer which solves the problem.

1. Use `fdisk -l SD.img` to list the partition table of our image. It will render this (sorry my system is not in english):
```
Disco SD.img: 210,02 MiB, 220217856 bytes, 430113 sectores
Unidades: sectores de 1 * 512 = 512 bytes 
Tamaño de sector (lógico/físico): 512 bytes / 512 bytes
Tamaño de E/S (mínimo/óptimo): 512 bytes / 512 bytes
Tipo de etiqueta de disco: gpt
Identificador del disco: E9E74BE8-F67F-4473-962F-035298072EEC

Dispositivo Comienzo  Final Sectores Tamaño Tipo
SD.img1        32768  65535    32768    16M Datos básicos de Microsoft
SD.img2        65536 430079   364544   178M Sistema EFI
```
2. We need to keep in mind two data from the previous output: **The sector size** (512 bytes) **and the end of the partition** (in our case the partition 2 ends at sector 4300079.
3. With those data in mind we will _shave_ the image file with `truncate`. But because we have a GPT partition table, we must take that into account and add a certain amount of free space at the end (33 sectors)
```
truncate --size=$[(430079+1+33)*512] SD.img <-- Notice the 33 extra sectors for GPT table backup
```
4. The problem now is the partition table is corrupted, because it looks like with GPT there is a copy of the partition table at the very end of the space and we just shaved it off, so we have to fix it. [You can find a detailed description of the process here](https://askubuntu.com/questions/1189682/how-to-truncate-a-disk-image-file-of-unused-space-without-corrupting-gpt-partiti), but to summarize you'll have to use gdisk to fix it. To do so, you just:

1. Verify the image file with `v`
2. Enter into expert mode with `x`
3. Relocate backup data structures to the end of the disk with `e`
4. Write changes to the image file with `w`

I've left the dump of the process [in this gist](https://gist.github.com/jtornero/8ea9990962b5d2ca8c4f277dd3b98d23).

**¡Done!** Now you have your tiny boot image (in my case just about 210 Mb) that is easy and fast to be written to your Rock 4C+ SD cards.
